File: 백준/Gold/15685. 드래곤 커브/드래곤 커브.py
import sys
input = sys.stdin.readline

n = int(input())
graph = [[0] * 101 for _ in range(101)]
# 0, 1, 2, 3 방향에 맞게 dx, dy 설정
dx = [0, -1, 0, 1]
dy = [1, 0, -1, 0]

for i in range(n):
    y, x, d, g = map(int, input().split())
    graph[x][y] = 1

    # 좌표가 나아갈 방향 리스트
    dir = [d]
    # 세대 수 만큼 반복
    for j in range(g):
        # 방향은 결국엔 시계방향으로 90도씩 돌기 때문에 (dir+1) % 4를 해준다.
        # 다음 세대로 나아갈 때는 출발해서 도착할때까지의 방향의 역순으로 배치가 된다.
        for k in range((len(dir))-1,-1,-1):
            dir.append((dir[k] + 1) % 4)

    for j in range(len(dir)):
        x+=dx[dir[j]]
        y+=dy[dir[j]]

        if x < 0 or x >= 101 or y < 0 or y >= 101:
            continue

        graph[x][y] = 1

ans = 0
for i in range(100):
    for j in range(100):
        if graph[i][j] == 1 and graph[i + 1][j] == 1 and graph[i][j + 1] == 1 and graph[i + 1][j + 1] == 1:
            ans += 1

print(ans)

File: 백준/Gold/15685. 드래곤 커브/드래곤 커브.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int[][] graph = new int[101][101];
    static int[] dx = {0, -1, 0, 1};
    static int[] dy = {1, 0, -1, 0};
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            int y = Integer.parseInt(st.nextToken());
            int x = Integer.parseInt(st.nextToken());
            int d = Integer.parseInt(st.nextToken());
            int g = Integer.parseInt(st.nextToken());
            graph[x][y] = 1;

            ArrayList<Integer> dir = new ArrayList<>();
            dir.add(d);

            // 세대 수 만큼 반복
            makeCurveDirection(dir, g);

            // 커브 좌표 체크
            checkCurveCoordinate(dir, x, y);
        }
        // 결과 계산
        int ans = calculateAnswer();
        sb.append(ans);
        System.out.println(sb);
    }

    static void checkCurveCoordinate(ArrayList<Integer> dir, int x, int y) {
        for (Integer integer : dir) {
            x += dx[integer];
            y += dy[integer];

            if (x < 0 || y < 0 || x >= 101 || y >= 101) {
                break;
            }

            graph[x][y] = 1;
        }
    }

    static void makeCurveDirection(ArrayList<Integer> dir, int g) {
        for (int i = 0; i < g; i++) {
            // 제일 끝쪽부터 새로운 커브가 만들어짐
            for (int j = dir.size() - 1; j > -1; j--) {
                dir.add((dir.get(j) + 1) % 4);
            }
        }
    }

    static int calculateAnswer() {
        int ans = 0;
        for (int i = 0; i < 100; i++) {
            for (int j = 0; j < 100; j++) {
                if (graph[i][j] == 1 && graph[i + 1][j] == 1 && graph[i][j + 1] == 1 && graph[i + 1][j + 1] == 1) {
                    ans++;
                }
            }
        }
        return ans;
    }
}